include/libcyphal/transport/can/media.hpp

```diff
@@ -0,0 +1,88 @@
+/// @copyright
+/// Copyright (C) OpenCyphal Development Team  <opencyphal.org>
+/// Copyright Amazon.com Inc. or its affiliates.
+/// SPDX-License-Identifier: MIT
+
+#ifndef LIBCYPHAL_TRANSPORT_CAN_MEDIA_HPP_INCLUDED
+#define LIBCYPHAL_TRANSPORT_CAN_MEDIA_HPP_INCLUDED
+
+#include "libcyphal/types.hpp"
+#include "libcyphal/transport/errors.hpp"
+#include "libcyphal/transport/defines.hpp"
+
+namespace libcyphal
+{
+namespace transport
+{
+namespace can
+{
+
+using CanId = std::uint32_t;
+
+struct Filter final
+{
+    CanId id;
+    CanId mask;
+};
+using Filters = cetl::span<const Filter>;
+
+struct RxMetadata final
+{
+    TimePoint   timestamp;
+    CanId       can_id;
+    std::size_t payload_size;
+};
+
+class IMedia
```

Comment:
We're missing a key feature I wanted to have where we implement "borrowed memory" that we ask for from the media layer and use that to write frames into. This enables implementations using DMA buffers where the peripheral TX buffer can be borrowed and used directly by our message serializers thus enabling no-copy serialization of data structures to be transmitted directly by hardware. It's easy enough to simply provide regular memory buffers and copy into sockets when implementing a media layer for something like linux but the borrowed-memory model allows MCUs to have incredibly efficient integrations when needed.  I want this to be a common pattern used by all media layer implementations regardless of transport.

One way to implement this is [DirectX style](https://learn.microsoft.com/en-us/windows/win32/api/d3d9helper/nf-d3d9helper-idirect3dvertexbuffer9-lock) where the act of successfully locking memory provides the caller a pointer they can use until they unlock. This works well for DirectX since the act of unlocking becomes an event the 3d engine uses to release the memory to the GPU.

For libcyphal, I'd rather find a way to implement a "media allocator" that is a polymorphic allocator obtained from the media layer and populated with a `memory_resource` implemented in terms of DMA, DTCM, plain-old-system memory, etc (i.e. the simplest linux media layer would simply use the system's new_delete_resource). The challenge with this is there is no "I'm done with this, please submit to hardware" event that is intrinsic to the mechanism. Overloading the deallocate method to do this seem ill-advised (i.e. "to send your message, delete it" is not really a good plan). We could actually subclass polymorphic_allocator and add lock/unlock region methods to it?

Reply:
This is not going to be an easy change at the moment because it affects the lizards. The transmission pipelines in both libudpard and libcanard are currently copying. You give the library your transfer payload (which currently has to be a contiguous buffer but it is quite easy to support scattered buffers which yields certain benefits), it allocates a sequence of frame payload buffers and copies fragments of your original buffer into them, adding frame headers/tails in the process. It is possible to eliminate the copying if we could allow the enqueued tx frame items (see `UdpardTxItem`, `CanardTxQueueItem`) to reference memory in the original transfer payload instead of containing own copes of it; said tx queue item objects will still need to be heap-allocated, though, but they would become very small.

To support this, the media layer implementations would have to be able to accept not just contiguous frame payloads (currently we push to the media layer complete buffers ready to be written into the socket or whatever) but fragments of them, so something similar to POSIX iov will be needed.

Most importantly, LibCyphal will need to keep the transfer payload buffer alive until all frames that originated from it have cleared the tx queue (either by being transmitted or by timing out). This is actually not super hard to implement because 1. Lizards provide continuity between the transfer pushed by the user and the frames that originate from it (see the `user_transfer_reference` field); and 2. Lizards return the number of frames enqueued per transfer; so we can implement reference counting quite easily.

Earlier, I mentioned that scattered transfer payload buffers offer "certain benefits" during transmission; what I mean is that if we support scattered buffers, DSDL serialization could be sped up considerably in cases where there are large byte arrays present in the message (like imagery data) because the serializer would be able to serialize only non-trivial parts of the message and keep references to the blobs:

```python
uavcan.time.SynchronizedTimestamp.1.0 timestamp  # NON ZERO-COPY
uint16 IMAGE_WIDTH  = 1280
uint16 IMAGE_HEIGHT = 800
uint8 BITS_PER_PIXEL = 24
uint8[IMAGE_WIDTH * IMAGE_HEIGHT * BITS_PER_PIXEL / 8] pixel_data  # ZERO-COPY!
```

If we go this way, to support full end-to-end low-copy operation, the application would need to allocate the message object itself from the media memory resource.

One side effect of scattered input buffers is that an output frame may end up straddling multiple fragments; to support this we will need to allow a tx queue item to reference not just one but an arbitrary number of fragments from the input buffer.

Am I thinking in the correct direction here?

Note that there are related considerations affecting the RX pipeline:

<img width="767" alt="image" src="https://github.com/OpenCyphal-Garage/libcyphal/assets/3298404/c5f5dfb8-8f2c-4e8d-b7ef-5cd0c172782b">

Reply:
Does it have to affect the lizards immediately? My thought is, as long as we establish that the media layer supplies memory which the higher layers are simply borrowing we are free to implement this against the lizards in various ways (the simplest being the current copy into their memory space we are setting up in our media layer apis just behind the media abstraction). This defers such optimizations to the underlying implementation of the media layer and supports various schemes. Am I being naive here?

Reply:
Moved into https://github.com/OpenCyphal-Garage/libcyphal/issues/352